<a href="https://www.bigdatauniversity.com"><img src="https://ibm.box.com/shared/static/cw2c7r3o20w9zn8gkecaeyjhgw3xdgbj.png" width="400" align="center"></a>

<h1 align=center><font size="5"> Coursework for week 3 of IBM Data Science course </font></h1>

In this notebook homework assignment for week 3 is completed, which is: __"Build the code to scrape the Wikipedia page "List of postal codes of Canada: M"__ , and after that explore and cluster the neighborhoods in Toronto.

The tutor's instructions are: *Use the Notebook to build the code to scrape the [Wikipedia page,](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M), in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below: (...)*

### __This notebook is prepared by Vladimir Belobragin.__
Please, follow the last version of the notebook in Github repository at: https://github.com/vbelobragin/Coursera_Capstone.git 

<h3 align=left><font size="3"> 1.Scrape the Wikipedia page "List of postal codes of Canada: M" </font></h3>


At first, import libraries.

In [187]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

from bs4 import BeautifulSoup # import recommended html parsing library

print('Libraries imported.')

Libraries imported.


Install visualization library, Folium, to visualize the results:

In [188]:
!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
geopy                     1.18.1                     py_0    conda-forge
Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda/envs/DSX-Python35:
#
folium                    0.5.0                      py_0    conda-forge


Create the data drame object:

In [189]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(source, 'lxml')
article = soup.find('table', class_='wikitable sortable')

table_list = []
for rows in article.find_all('td'):
    row = rows.text
    row = row.replace('\n', '')
    table_list.append(row) 
           


Define the dataframe object:

In [190]:
column_names = ['PostalCode', 'Borough', 'Neighborhood'] 
df=[]
df=pd.DataFrame(columns=column_names)

Paste the codes to the dataframe and drop non-assighned boroughs:

In [191]:
df.iloc[:,0]=table_list[::3]
df.iloc[:,1]=table_list[1::3]
df.iloc[:,2]=table_list[2::3]
df.replace("Not assigned", np.nan, inplace = True)
df.dropna(subset=["Borough"], axis=0, inplace = True)
df.reset_index(drop=True, inplace=True)

Now assign to the non-assighned neighborhoods names of boroughs:

In [192]:
for i in range(0, df.shape[0]):
    if pd.isnull(df.loc[i,'Neighborhood']):
        df.replace(df.loc[i,'Neighborhood'], df.loc[i,'Borough'],inplace=True)
df.head(15)

PostalCode           Borough      Neighborhood
0         M3A        North York         Parkwoods
1         M4A        North York  Victoria Village
2         M5A  Downtown Toronto      Harbourfront
3         M5A  Downtown Toronto       Regent Park
4         M6A        North York  Lawrence Heights
5         M6A        North York    Lawrence Manor
6         M7A      Queen's Park      Queen's Park
7         M9A         Etobicoke  Islington Avenue
8         M1B       Scarborough             Rouge
9         M1B       Scarborough           Malvern
10        M3B        North York   Don Mills North
11        M4B         East York  Woodbine Gardens
12        M4B         East York     Parkview Hill
13        M5B  Downtown Toronto           Ryerson
14        M5B  Downtown Toronto   Garden District

Combine rows with the same code into one row and separate them with a comma:

In [193]:
df['Neighborhood'] = df[['PostalCode','Borough','Neighborhood']].groupby(['PostalCode','Borough'])['Neighborhood'].transform(lambda x: ','.join(x)) 
df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)

df.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                            Harbourfront,Regent Park  
3                     Lawrence Heights,Lawrence Manor  
4                                        Queen's Park  
5                                    Islington Avenue  
6                                       Rouge,Malvern  
7                                     Don Mills North  
8                      Woodbine Gardens,Parkview Hill  
9                             Ryerson,Garden District  
10                                          Glencairn  
11  Cloverdale,Islington,Martin Grove,Princess Gar...  
12               Highland Creek,Rouge Hill,Port Union  
13                    Flemingdon Park,Don Mills South  
14                                   Woodbine Heights

Now get the final shape of the DataFrame:

In [194]:
df.shape

(103, 3)

<h3 align=left><font size="3"> 2. Load a csv file that has the geographical coordinates of each postal code </font></h3>

Given that package Geocoder is DEFINITELY very unreliable, use a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

In [195]:
TorontoCoord=df
TorontoCoord1=pd.read_csv("http://cocl.us/Geospatial_data")
TorontoCoord[['Latitude', 'Longitude']]=TorontoCoord1[['Latitude', 'Longitude']]
TorontoCoord.head()

PostalCode           Borough                     Neighborhood   Latitude  \
0        M3A        North York                        Parkwoods  43.806686   
1        M4A        North York                 Victoria Village  43.784535   
2        M5A  Downtown Toronto         Harbourfront,Regent Park  43.763573   
3        M6A        North York  Lawrence Heights,Lawrence Manor  43.770992   
4        M7A      Queen's Park                     Queen's Park  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [196]:
TorontoCoord.shape

(103, 5)

<h3 align=left><font size="3"> 3. Map of neighborhoods in Toronto.  </font></h3> 

Get the geographical coordinates of Toronto:

In [197]:
address = 'Toronto Canada'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Create a map of Toronto with neighborhoods superimposed on top:

In [198]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(TorontoCoord['Latitude'], TorontoCoord['Longitude'], TorontoCoord['Borough'], TorontoCoord['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<h3 align=left><font size="3"> 4. Get the list of venues in Toronto.  </font></h3> 

Unfortunately (maybe, FORTUNATELY) Toronto is not as big city as New York. There are only 103 Boroughs in Toronto. Moreover, we do not have coordinates of Neighborhoods in Toronto, only lattitude and longitude of the Postal Codes zones. So, we can only cluster neighborhoods based on the postal codes.

<h3 align=left><font size="2"> Define Foursquare Credentials and Version </font></h3>

__Please be aware, that for privacy purpose variable CLIENT_SECRET is omitted from the GitHub version of the notebook.__ 

In [199]:
CLIENT_ID = 'HSB35PAX0VVRGDIW042W2DIQZ53JRLS2GF3O4X4Y3QWPP5RI' # your Foursquare ID
CLIENT_SECRET = '**********' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET: **********')

Your credentails:
CLIENT_ID: HSB35PAX0VVRGDIW042W2DIQZ53JRLS2GF3O4X4Y3QWPP5RI
CLIENT_SECRET: **********


Borrow the get_category_type function from the Foursquare lab.

In [200]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    

Create a function to repeat the process to get the top 50 venues that are in every neighborhood within a radius of 500 meters - for all the neighborhoods in Toronto (based on post codes):

In [201]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
                    
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called Toronto_venues.

In [202]:
Toronto_venues = getNearbyVenues(names=TorontoCoord['Neighborhood'],
                                   latitudes=TorontoCoord['Latitude'],
                                   longitudes=TorontoCoord['Longitude']
                                  )

Parkwoods
Victoria Village
Harbourfront,Regent Park
Lawrence Heights,Lawrence Manor
Queen's Park
Islington Avenue
Rouge,Malvern
Don Mills North
Woodbine Gardens,Parkview Hill
Ryerson,Garden District
Glencairn
Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park
Highland Creek,Rouge Hill,Port Union
Flemingdon Park,Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe
Guildwood,Morningside,West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,Downsview North,Wilson Heights
Thorncliffe Park
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Scarborough Village
Fairview,Henry Farm,Oriole
Northwood Park,York University
East Toronto
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
East Birchmount Park,Ionview,Kennedy Park
Bayview Village
CFB Toronto,Downsview East
The Danforth West,Riverdale
Design E

In [203]:
print(Toronto_venues.shape)
print('Note: there are {} uniques categories of Neighborhood.'.format(len(Toronto_venues['Neighborhood'].unique())))
Toronto_venues.head()


(2243, 7)
Note: there are 101 uniques categories of Neighborhood.


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                 Parkwoods              43.806686              -79.194353   
1          Victoria Village              43.784535              -79.160497   
2          Victoria Village              43.784535              -79.160497   
3  Harbourfront,Regent Park              43.763573              -79.188711   
4  Harbourfront,Regent Park              43.763573              -79.188711   

                             Venue  Venue Latitude  Venue Longitude  \
0                          Wendy's       43.807448       -79.199056   
1            Royal Canadian Legion       43.782533       -79.163085   
2   Scarborough Historical Society       43.788755       -79.162438   
3  Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
4                G & G Electronics       43.765309       -79.191537   

         Venue Category  
0  Fast Food Restaurant  
1                   Bar  
2        History Museum  
3           Pizza Place  
4     Electronics Store

As we see, we got 1697 venues, which is a reasonable number of venues for clustering purpose.

Let's check how many venues were returned for each neighborhood

In [204]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                  3   
Agincourt                                                              19   
Agincourt North,L'Amoreaux East,Milliken,Steele...                     42   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                     10   
Bathurst Manor,Downsview North,Wilson Heights                          16   
Bayview Village                                                        17   
Bedford Park,Lawrence Manor East                                      100   
Berczy Park                                                             2   
Birch Cliff,Cliffside West                                            100   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      5   
Brockton,Exhibition Place,Parkdale Village                             39   
Business Reply Mail Processing Centre 969 Eastern                       4   
CFB Toronto,Downsview East                                              3   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     18   
Cabbagetown,St. James Town                                              1   
Caledonia-Fairbanks                                                     1   
Canada Post Gateway Processing Centre                                  19   
Cedarbrae                                                               5   
Central Bay Street                                                      5   
Chinatown,Grange Park,Kensington Market                                38   
Christie                                                                4   
Church and Wellesley                                                    6   
Clairlea,Golden Mile,Oakridge                                           4   
Clarks Corners,Sullivan,Tam O'Shanter                                  23   
Cliffcrest,Cliffside,Scarborough Village West                          48   
Cloverdale,Islington,Martin Grove,Princess Gard...                      7   
Commerce Court,Victoria Hotel                                           1   
Davisville                                                              4   
Davisville North                                                      100   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     11   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                            56   
Design Exchange,Toronto Dominion Centre                                20   
Don Mills North                                                        11   
Dorset Park,Scarborough Town Centre,Wexford Hei...                     25   
Dovercourt Village,Dufferin                                             4   
Downsview Central                                                      48   
Downsview Northwest                                                   100   
Downsview West                                                         18   
East Birchmount Park,Ionview,Kennedy Park                              32   
East Toronto                                                           13   
Emery,Humberlea                                                        82   
Fairview,Henry Farm,Oriole                                              4   
First Canadian Place,Underground city                                   2   
Flemingdon Park,Don Mills South                                         9   
Forest Hill North,Forest Hill West                                     14   
Glencairn                                                               6   
Guildwood,Morningside,West Hill                                        61   
Harbord,University of Toronto                                           4   
Harbourfront East,Toronto Islands,Union Station                         9   
Harbourfront,Regent Park                                                7   
High Park,The Junction South                    

__Let's find out how many unique categories can be curated from all the returned venues__

In [205]:
Toronto_venues[Toronto_venues['Venue Category']=='Neighborhood']

Neighborhood  Neighborhood Latitude  \
303                      Little Portugal,Trinity              43.676357   
427   Brockton,Exhibition Place,Parkdale Village              43.659526   
1079                  Birch Cliff,Cliffside West              43.650571   
1166                            Willowdale South              43.640816   

      Neighborhood Longitude             Venue  Venue Latitude  \
303               -79.293031     Upper Beaches       43.680563   
427               -79.340923       Leslieville       43.662070   
1079              -79.384568  Downtown Toronto       43.653232   
1166              -79.381752      Harbourfront       43.639526   

      Venue Longitude Venue Category  
303        -79.292869   Neighborhood  
427        -79.337856   Neighborhood  
1079       -79.385296   Neighborhood  
1166       -79.380688   Neighborhood

In [206]:
Toronto_venues.loc[{292,416,893,930},'Venue Category']='Neighborhood1'
#Toronto_venues[Toronto_venues['Venue Category']=='Neighborhood1']

<h3 align=left><font size="3"> 4. Analyze Neighborhoods in Toronto.  </font></h3> 

In [207]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move _Neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head(10)

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   
5            0                  0               0                  0        0   
6            0                  0               0                  0        0   
7            0                  0               0                  0        0   
8            0                  0               0                  0        0   
9            0                  0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   
5                   0             0               0                0   
6                   0             0               0                0   
7                   0             0               0                0   
8                   0             0               0                0   
9                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   
5                 0                    0             0         0   
6                 0                    0             0         0   
7                 0                    0             0         0   
8                 0                    0             0         0   
9                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   
5                 0            0           0                    0   
6                 0            0           0                    0   
7                 0            0           0                    0   
8                 0            0           0                    0   
9                 0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   
5                 0                   0            0              0   
6                 0                   0            0              0   
7                 0                   0            0              0   
8                 0                   0            0              0   
9                 0                   0            0              0   

   BBQ Joint  Baby Store  B

And let's examine the new dataframe size.

In [208]:
Toronto_onehot.shape

(2243, 280)

__Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category__

In [209]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                             Adelaide,King,Richmond      0.00000   
1                                          Agincourt      0.00000   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...      0.02381   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...      0.00000   
4      Bathurst Manor,Downsview North,Wilson Heights      0.00000   

   Accessories Store  Adult Boutique  Afghan Restaurant   Airport  \
0                0.0             0.0                0.0  0.333333   
1                0.0             0.0                0.0  0.000000   
2                0.0             0.0                0.0  0.000000   
3                0.0             0.0                0.0  0.000000   
4                0.0             0.0                0.0  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0               0.0                  0.0           0.0       0.0   
1               0.0                  0.0           0.0       0.0   
2               0.0                  0.0           0.0       0.0   
3               0.0                  0.0           0.0       0.0   
4               0.0                  0.0           0.0       0.0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0               0.0          0.0         0.0              0.00000   
1               0.0          0.0         0.0              0.00000   
2               0.0          0.0         0.0              0.02381   
3               0.0          0.0         0.0              0.00000   
4               0.0          0.0         0.0              0.00000   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0               0.0                 0.0          0.0            0.0   
1               0.0                 0.0          0.0            0.0   
2               0.0                 0.0          0.0            0.0   
3               0.0                 0.1          0.0            0.0   
4               0.0                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop    Bakery    Bank       Bar  \
0        0.0         0.0         0.0  0.000000  0.0000  0.000000   
1        0.0         0.0         0.0  0.052632  0.0000  0.052632   
2        0.0         0.0         0.0  0.000000  0.0000  0.023810   
3        0.0         0.0         0.0  0.000000  0.0000  0.000000   
4        0.0         0.0         0.0  0.000000  0.0625  0.000000   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Beach  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  \
0    0.0       0.0         0.0                 0.0        0.0     0.0   
1    0.0       0.0         0.0                 0.0        0.0     0.0   
2    0.0       0.0         0.0                 0.0        0.0     0.0   
3    0.0       0.0         0.0                 0.0        0.0     0.0   
4    0.0       0.0         0.0                 0.0        0.0     0.0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0         0.0            0.0        0.0       0.0                   0.0   
1         0.0            0.0        0.0       0.0                

__Let's confirm the new size__

In [210]:
Toronto_grouped.shape

(101, 280)

__Let's print each neighborhood along with the top 5 most common venues__

In [211]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    #print(temp)
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')
   

----Adelaide,King,Richmond----
                venue  freq
0            Bus Stop  0.33
1             Airport  0.33
2                Park  0.33
3         Yoga Studio  0.00
4  Miscellaneous Shop  0.00


----Agincourt----
            venue  freq
0  Breakfast Spot  0.11
1     Coffee Shop  0.11
2            Café  0.11
3         Stadium  0.05
4          Bakery  0.05


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                 venue  freq
0          Coffee Shop  0.24
1                  Gym  0.05
2                Diner  0.05
3  Japanese Restaurant  0.05
4     Sushi Restaurant  0.05


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
          venue  freq
0   Pizza Place   0.2
1      Pharmacy   0.1
2  Skating Rink   0.1
3           Gym   0.1
4   Coffee Shop   0.1


----Bathurst Manor,Downsview North,Wilson Heights----
                venue  freq
0         Coffee Shop  0.12
1       Deli / Bodega  0.06
2             

__Let's put that into a pandas dataframe__

First, let's write a function to sort the venues in descending order.

In [212]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [213]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()


Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond                  Park   
1                                          Agincourt           Coffee Shop   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...           Coffee Shop   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...           Pizza Place   
4      Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0               Airport              Bus Stop           Event Space   
1        Breakfast Spot                  Café          Climbing Gym   
2      Sushi Restaurant                 Diner                   Gym   
3                   Gym    Athletics & Sports              Pharmacy   
4    Frozen Yogurt Shop  Fast Food Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Ethiopian Restaurant    Falafel Restaurant   Empanada Restaurant   
1                   Bar    Italian Restaurant               Stadium   
2   Japanese Restaurant                  Café                   Bar   
3                  Pool                   Pub        Sandwich Place   
4           Bridal Shop                 Diner                  Bank   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0      Electronics Store  Eastern European Restaurant           Dessert Shop  
1   Caribbean Restaurant       Furniture / Home Store          Burrito Place  
2  Portuguese Restaurant                          Spa     Italian Restaurant  
3           Skating Rink                  Coffee Shop         General Travel  
4             Restaurant                Deli / Bodega       Sushi Restaurant

<h3 align=left><font size="3"> 5. Cluster Neighborhoods.  </font></h3> 

Run k-means to cluster the neighborhood into 6 clusters.

In [214]:
# set number of clusters
kclusters = 4

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100]


array([2, 3, 3, 3, 3, 3, 3, 2, 3, 3, 3, 2, 2, 3, 3, 3, 3, 3, 3, 3, 2, 3, 2,
       3, 3, 3, 1, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 1, 3, 3, 2, 3, 3, 2, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 2, 3, 0, 3, 3, 3, 3, 0, 3, 3, 3, 3, 2, 3, 3, 3, 2, 2, 3, 3,
       3, 3, 2, 3, 2, 3, 3, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [215]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = TorontoCoord

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
Toronto_merged.dropna(subset=["Cluster Labels"], axis=0, inplace = True)
Toronto_merged['Cluster Labels'] = Toronto_merged['Cluster Labels'].astype(int)

Toronto_merged.describe()
#Toronto_merged.head() # check the last columns!

Latitude   Longitude  Cluster Labels
count  101.000000  101.000000      101.000000
mean    43.703669  -79.397712        2.693069
std      0.051186    0.095271        0.674427
min     43.602414  -79.615819        0.000000
25%     43.659526  -79.464763        3.000000
50%     43.696948  -79.388790        3.000000
75%     43.744734  -79.340923        3.000000
max     43.815252  -79.160497        3.000000

Finally, let's visualize the resulting clusters

In [216]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<h3 align=left><font size="3"> 6. Examine Clusters.  </font></h3> 

Examine each cluster and determine the discriminating venue categories that distinguish each cluster. 

__Cluster 1__

In [217]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
32       Scarborough               0              Korean Restaurant   
91  Downtown Toronto               0                 Baseball Field   
97  Downtown Toronto               0  Paper / Office Supplies Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32        Baseball Field            Food Truck          Home Service   
91         Women's Store   Dumpling Restaurant                 Diner   
97        Baseball Field         Women's Store   Dumpling Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
32        Discount Store               Dog Run      Doner Restaurant   
91        Discount Store               Dog Run      Doner Restaurant   
97                 Diner        Discount Store               Dog Run   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
32            Donut Shop             Drugstore          Dumpling Restaurant  
91            Donut Shop             Drugstore  Eastern European Restaurant  
97      Doner Restaurant            Donut Shop                    Drugstore

This is definetly, a baseball funs cluster...

__Cluster 2__

In [218]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
5          Etobicoke               1     Convenience Store   
14         East York               1                  Park   
48  Downtown Toronto               1            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5             Playground   Dumpling Restaurant                 Diner   
14            Playground            Donut Shop          Dessert Shop   
48         Women's Store             Drugstore    Dim Sum Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5         Discount Store               Dog Run      Doner Restaurant   
14    Dim Sum Restaurant                 Diner        Discount Store   
48                 Diner        Discount Store               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5             Donut Shop             Drugstore          Women's Store  
14               Dog Run      Doner Restaurant              Drugstore  
48      Doner Restaurant            Donut Shop    Dumpling Restaurant

This is a playground cluster. Nice for children!

__Cluster 3__

In [219]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York               2  Fast Food Restaurant   
20   Downtown Toronto               2                  Park   
23          East York               2                  Park   
25   Downtown Toronto               2  Fast Food Restaurant   
30   Downtown Toronto               2                  Park   
40         North York               2                  Park   
44        Scarborough               2                  Park   
50         North York               2                  Park   
63               York               2                Garden   
64               York               2         Jewelry Store   
72         North York               2                  Park   
74    Central Toronto               2                  Park   
79    Central Toronto               2                  Park   
90        Scarborough               2                  Park   
98          Etobicoke               2                  Park   
100      East Toronto               2                  Park   

    2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0           Women's Store                   Drugstore    Dim Sum Restaurant   
20              Cafeteria            Department Store    Dim Sum Restaurant   
23                   Bank                       Diner        Discount Store   
25      Food & Drink Shop                    Bus Stop                  Park   
30                Airport                    Bus Stop           Event Space   
40            Coffee Shop           Convenience Store    Falafel Restaurant   
44               Bus Line                 Swim School    Dim Sum Restaurant   
50                  Trail                  Playground          Dessert Shop   
63              Drugstore          Dim Sum Restaurant                 Diner   
64                  Trail                        Park      Sushi Restaurant   
72    Japanese Restaurant                         Pub                Bakery   
74   Fast Food Restaurant                    Pharmacy                Market   
79       Basketball Court  Construction & Landscaping                Bakery   
90                  River                        Pool          Dessert Shop   
98      Convenience Store         Dumpling Restaurant                 Diner   
100              Bus Line           Mobile Phone Shop           Pizza Place   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Diner        Discount Store               Dog Run   
20                  Diner        Discount Store               Dog Run   
23                Dog Run      Doner Restaurant            Donut Shop   
25                    Gym      Department Store  Ethiopian Restaurant   
30   Ethiopian Restaurant    Falafel Restaurant   Empanada Restaurant   
40            Event Space  Ethiopian Restaurant   Empanada Restaurant   
44         Farmers Market                 Diner        Discount Store   
50     Dim Sum Restaurant                 Diner        Discount Store   
63         Discount Store               Dog Run      Doner Restaurant   
64     Dim Sum Restaurant                 Diner        Discount Store   
72          Women's Store            Donut Shop                 Diner   
74          Women's Store      Greek Restaurant          Dance Studio   
79    Dumpling Restaurant        Discount Store               Dog Run   
90     Dim Sum Restaurant                 Diner        Discount Store   
98         Discount Store               Dog Run      Doner Restaurant   
100            Donut Shop                 Diner        Discount Store   

    8th Most Common Venue        9th Most Common Venue  \
0        Doner Restaurant                   Donut Shop   
20       Doner Restaurant                   Donut Shop   
23              Drugstore          Dumpling Restaurant   
25    Empanada Restaurant            Electronics Store   
30      Electronics Store  Eastern European Restaurant   
40      

__Cluster 4__

In [220]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1          North York               3                        Bar   
2    Downtown Toronto               3             Breakfast Spot   
3          North York               3                Coffee Shop   
4        Queen's Park               3                     Lounge   
6         Scarborough               3             Discount Store   
7          North York               3                   Bus Line   
8           East York               3              Movie Theater   
9    Downtown Toronto               3            College Stadium   
10         North York               3          Indian Restaurant   
11          Etobicoke               3  Middle Eastern Restaurant   
12        Scarborough               3                     Lounge   
13         North York               3                Pizza Place   
15   Downtown Toronto               3       Fast Food Restaurant   
17          Etobicoke               3       Fast Food Restaurant   
18        Scarborough               3             Clothing Store   
19       East Toronto               3         Chinese Restaurant   
21               York               3                        Gym   
22        Scarborough               3                 Restaurant   
24   Downtown Toronto               3                Coffee Shop   
26        Scarborough               3       Gym / Fitness Center   
27         North York               3                Coffee Shop   
28         North York               3                Coffee Shop   
29          East York               3     Furniture / Home Store   
31       West Toronto               3                       Bank   
33         North York               3               Liquor Store   
34         North York               3                Coffee Shop   
35          East York               3                Pizza Place   
36   Downtown Toronto               3                Curling Ice   
37       West Toronto               3                Coffee Shop   
38        Scarborough               3                Coffee Shop   
39         North York               3          Indian Restaurant   
41       East Toronto               3           Greek Restaurant   
42   Downtown Toronto               3           Sushi Restaurant   
43       West Toronto               3                       Café   
45         North York               3                       Park   
46         North York               3             Clothing Store   
47       East Toronto               3             Sandwich Place   
49         North York               3                Coffee Shop   
51        Scarborough               3                Coffee Shop   
52         North York               3        Japanese Restaurant   
53         North York               3                Coffee Shop   
54       East Toronto               3             Clothing Store   
55         North York               3                Coffee Shop   
56               York               3                Coffee Shop   
57         North York               3                Coffee Shop   
58        Scarborough               3                Coffee Shop   
59         North York               3                Coffee Shop   
60         North York               3                Coffee Shop   
61    Central Toronto               3                Coffee Shop   
62    Central Toronto               3                Coffee Shop   
65        Scarborough               3                       Café   
66         North York               3                       Café   
67    Central Toronto               3                       Café   
68    Central Toronto               3             Airport Lounge   
69       West Toronto               3                Coffee Shop   
70          Etobicoke               3                       Café   
71        Scarborough               3             Clothing Store   
73    Central Toronto               3                      Field   
75

### THANK YOU.

In [ ]:
DisableHTML5Autoplay_Initialize